File: WEEK4/bj2_1707.py
import sys
from collections import deque

input = sys.stdin.readline
sys.setrecursionlimit(200000)


def bfs(start_node, v, adjList, groups):
    queue = deque([start_node])
    groups[start_node] = 1

    while queue:
        current = queue.popleft()
        next_group = 2 if groups[current] == 1 else 1

        for neighbor in adjList[current]:
            if groups[neighbor] == 0:
                groups[neighbor] = next_group
                queue.append(neighbor)
            elif groups[neighbor] == groups[current]:
                return False
    return True


k = int(input())
for _ in range(k):
    v, e = map(int, input().split())
    adjList = [[] for _ in range(v + 1)]
    groups = [0] * (v + 1)

    for _ in range(e):
        u, w = map(int, input().split())
        adjList[u].append(w)
        adjList[w].append(u)

    is_bipartite = True
    for i in range(1, v + 1):
        if groups[i] == 0:
            if not bfs(i, v, adjList, groups):
                is_bipartite = False
                break

    print("YES" if is_bipartite else "NO")

File: WEEK4/bj2_1707.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

let lineIndex = 0;
const k = Number(input[lineIndex++]);
const GROUP_A = 1;
const GROUP_B = 2;
let result = '';

function isBipartite(startNode, v, adjList, groups) {
    const queue = [];
    let head = 0;
    queue.push(startNode);
    groups[startNode] = GROUP_A;

    while (head < queue.length) {
        const current = queue[head++];
        const nextGroup = (groups[current] === GROUP_A) ? GROUP_B : GROUP_A;

        for (const neighbor of adjList[current]) {
            if (groups[neighbor] === 0) {
                groups[neighbor] = nextGroup;
                queue.push(neighbor);
            } else if (groups[neighbor] === groups[current]) {
                return false;
            }
        }
    }
    return true;
}

for (let t = 0; t < k; t++) {
    const [v, e] = input[lineIndex++].split(' ').map(Number);
    const adjList = Array.from({ length: v + 1 }, () => []);
    const groups = new Array(v + 1).fill(0);

    for (let i = 0; i < e; i++) {
        const [u, w] = input[lineIndex++].split(' ').map(Number);
        adjList[u].push(w);
        adjList[w].push(u);
    }

    let isBipartiteGraph = true;
    for (let i = 1; i <= v; i++) {
        if (groups[i] === 0) {
            if (!isBipartite(i, v, adjList, groups)) {
                isBipartiteGraph = false;
                break;
            }
        }
    }
    result += (isBipartiteGraph ? "YES" : "NO") + "\n";
}

console.log(result.trim());
